File: src/ch3/rand_xorshift.py
# Pythonで疑似乱数を生成する
import time

# 乱数生成のための種をグローバルで指定 --- (*1)
SEED = 0

# startからendの乱数を生成 --- (*2)
def rand(start, end):
    global SEED
    # 初回のみ乱数を初期化
    if SEED == 0:
        # 現在時刻で種を初期化
        SEED = int(time.time() * 1000)
    # 乱数を生成
    SEED ^= (SEED << 13) & 0xFFFFFFFF
    SEED ^= (SEED >> 17) & 0xFFFFFFFF
    SEED ^= (SEED << 5) & 0xFFFFFFFF
    return SEED % (end - start + 1) + start

# 100回乱数を表示 --- (*3)
for _ in range(100):
    print(rand(1, 6))

File: src/ch3/rand_xorshift.rs
use std::time::{SystemTime, UNIX_EPOCH};

fn main() {
    // 乱数を初期化 --- (*1)
    let mut seed = rand_init();
    // 100個の乱数を表示 --- (*2)
    for _ in 0..100 {
        // 乱数を生成 --- (*3)
        let v = rand(&mut seed, 1, 6);
        println!("{}", v);
    }
}

// 乱数を初期化する関数 --- (*4)
fn rand_init() -> u32 {
    // 現在時刻を利用して乱数を初期化
    SystemTime::now()
      .duration_since(UNIX_EPOCH).unwrap()
      .as_millis() as u32
}

// startからendの乱数を生成する関数 --- (*5)
fn rand(seed: &mut u32, start: u32, end: u32) -> u32 {
    *seed ^= *seed << 13;
    *seed ^= *seed >> 17;
    *seed ^= *seed << 5;
    return *seed % (end - start + 1) + start;
}


